File: pomade-signer/src/mailer/sendgrid.rs
use serde_json::json;

use crate::mailer::{Email, MailFuture, Mailer};

pub struct SendgridMailer {
    pub client: reqwest::Client,
    pub api_key: String,
}

impl Mailer for SendgridMailer {
    fn send(&self, from_email: &str, from_name: &str, email: Email) -> MailFuture {
        let client = self.client.clone();
        let key = self.api_key.clone();
        let from_email = from_email.to_string();
        let from_name = from_name.to_string();
        Box::pin(async move {
            let res = client
                .post("https://api.sendgrid.com/v3/mail/send")
                .bearer_auth(&key)
                .json(&json!({
                    "from": {"email": from_email, "name": from_name},
                    "personalizations": [{"to": [{"email": email.to}]}],
                    "subject": email.subject,
                    "content": [
                        {"type": "text/plain", "value": email.text},
                        {"type": "text/html", "value": email.html},
                    ],
                }))
                .send()
                .await
                .map_err(|e| e.to_string())?;

            if res.status().is_success() {
                Ok(())
            } else {
                Err(format!("sendgrid error: {}", res.status()))
            }
        })
    }
}

File: pomade-signer-go/mailer/sendgrid.go
package mailer

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

type SendgridMailer struct {
	Client *http.Client
	APIKey string
}

func (m SendgridMailer) Send(ctx context.Context, fromEmail string, fromName string, email Email) error {
	body, _ := json.Marshal(map[string]any{
		"personalizations": []any{map[string]any{"to": []any{map[string]any{"email": email.To}}}},
		"from":             map[string]any{"email": fromEmail, "name": fromName},
		"subject":          email.Subject,
		"content": []any{
			map[string]any{"type": "text/plain", "value": email.Text},
			map[string]any{"type": "text/html", "value": email.HTML},
		},
	})
	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.sendgrid.com/v3/mail/send", bytes.NewReader(body))
	req.Header.Set("Authorization", "Bearer "+m.APIKey)
	req.Header.Set("Content-Type", "application/json")
	res, err := m.Client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode > 299 {
		return fmt.Errorf("sendgrid error: %s", res.Status)
	}
	return nil
}
